### ISSUE TITLE
[Tracking] Redesign Unread Markers

### ISSUE THREAD

**Issue by JmillsExpensify** [Expensify/App#15212]
### Problem
We've struggled with on-going issues and regressions related to the reliability of the green "new" line that marks the boundary between new chats and old ones. At the moment we minimally have several issues that are being tackled separately, though the root cause is arguably the same. We're duplicating work, or worse, fixing things in one place to potentially break them in the other. 

### Solution
Let's devise a holistic plan for the "new" marker that definitely solves the root cause, and just as importantly, ensures that our solution creates stability around a feature that has been unstable so far. Knowing where a conversation ended and where it's starting anew is critical, because after all, we're a chat app. 

### Current Issues
- https://github.com/Expensify/App/issues/13740
- https://github.com/Expensify/App/issues/15019
- https://github.com/Expensify/App/issues/15064
- https://github.com/Expensify/App/issues/15266
- https://github.com/Expensify/App/issues/15742
- https://github.com/Expensify/App/issues/17207
- https://github.com/Expensify/App/issues/21288
- https://github.com/Expensify/App/issues/21453
- https://github.com/Expensify/App/issues/20959
- https://github.com/Expensify/App/issues/22238

--- 

# [DESIGN DOC ➡️](https://docs.google.com/document/d/144r92kq3lvJlT2sg9wQGzFHdq5PPyai4eCs5jsBWrZw/edit#heading=h.mnjguil21u0a)

## Proposal

Redesign Unread marker(The "New" marker line)

whatsnext: https://expensify.slack.com/archives/CC7NECV4L/p1686749939851689
Pre-design1: https://expensify.slack.com/archives/C01GTK53T8Q/p1678250769318789
Pre-design2: https://expensify.slack.com/archives/C01GTK53T8Q/p1684836449852569

## Tasks

- [x] Post **Proposal (full Problem/Solution statement)** in `#whatsnext`
- [x] Wait at least one full business day, and until the post has a majority (2/3) of positive reactions (👍)
- [x] Paste Proposal in the space above with a link to the Slack thread
- [x] Email `strategy@expensify.com` and paste in the Proposal
- [x] Fill out the **High-level overview of the problem**, **Timeline**, and **Terminology** sections of the Design Doc
- [x] Email `strategy@expensify.com` (continue the same email chain as before) with the link to your Design Doc
- [x] Host a **pre-design** meeting ([example](https://expensify.slack.com/archives/C01GTK53T8Q/p1665794669539419)) in an appropriate slack channel to discuss any necessary details in public before filling out the **High-level of proposed solution** section.
- [x] Fill out the **High-level of proposed solution** section
- [x] Email `stategy@expensify.com` again with links to the doc and pre-design conversation in Slack
- [x] Add the `DesignDocReview` label to get the **High-level of proposed solution** section reviewed
- [x] Respond to any questions or concerns and bring up blockers in Slack to get a consensus if necessary
- [x] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [x] Host another pre-design meeting in the appropriate slack channel to ask for engineering feedback on the technical solution.
- [x] Fill out the **Detailed implementation of the solution** and related sections.
- [x] Re-add the `DesignDocReview` label to this issue
- [x] Respond to any questions or concerns and bring up blockers in Slack to get consensus if necessary
- [x] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [x] Email `strategy@expensify.com` one last time to let them know the Design Doc is moving into the implementation phase
- [x] Implement the changes
- [x] Send out a follow up email to `strategy@expensify.com` once everything has been implemented and do a **Project Wrap-Up** retrospective that provides:
    - Summary of what we accomplished with this project
    - What went well?
    - What could we have done better?
    - What did we learn?

**Comment by bondydaa** [comment]
Just trying to jot down some notes for us while I was looking into a bit more about the `OpenReport` API endpoint.

It doesn't appear when we did some refactors that the `shouldMarkLastAsRead` param was added to really help with the "new message indicator".

The param was added in this PR https://github.com/Expensify/Web-Expensify/pull/34642 which was a part of some refactors for ReconnectReport API and was basically added so we could ensure we didn't mark things as read when reloading report data in `ReportActionsView`.

The other PR where it was slightly touched was here https://github.com/Expensify/Web-Expensify/pull/34931 which made it so that `OpenReport` would also create a chat if it didn't exist.

As discussed in this massive thread: https://expensify.slack.com/archives/C01GTK53T8Q/p1676267128690339

There are sort of 2 problems right now with the "new message indicator":
1. Whenever we call `OpenReport` directly we always are updating the last read NVP for a report so even if there are unread messages we blindly say "now" is the last time it was read and will basically mean all messages have been read.
2. There is some complex logic we use for trying to determine the "last read" report action that requires us to find the reportActionID and pass it down from the ReportActionView down to the individual ReportActionItem component. 

I'm not sure solving either of these individually will fix all of these so we should try address both now and get this functionality stable.

I think we should try to come at this problem from this point of view:
If this code didn't exist and we instead said "we need to write logic that ensures we highlight unread messages when the report is loaded" what would the solution be and does that align with what we currently have? (not that I think the current code is wrong I just want to make sure we aren't trying to force the solution to match what the code currently is)

I think the main question I think we should agree on is:
At what point should we mark a report or individual report action as "read"?

For me the more I dwell on it the more I think maybe we shouldn't update the last read when we load the report but after the user has take some action on the report or the component unmounts.

**Comment by chiragsalian** [comment]
Added https://github.com/Expensify/App/issues/15266 to the list too since the solution gets trickier than expected because of the new marker line.

**Comment by JmillsExpensify** [comment]
Nice, the more the merrier!

**Comment by chiragsalian** [comment]
So whats the plan here, are we focusing on this because I'm confused if we should be expanding the logic around the new marker logic ([example](https://github.com/Expensify/App/issues/15266#issuecomment-1439799125)) or should we pause on other issues until this one is done since the logic can change here?

**Comment by JmillsExpensify** [comment]
Good question. This issue was primarily meant to help us track all the issues that we're experiencing with the new marker line. I wouldn't pause on anything in particular, we just want to make sure that we don't have duplicates on our hands, and related that by grouping the issues together we can identify the root cause(s) more effectively.

**Comment by bondydaa** [comment]
was just sync'ing with @puneetlath about this. 

I think we should see about getting some of the expert contributors to help us get this done so I'm going to close out the individual issues linked on here so that we don't have it spreads across too many people getting bumped for updates and such and can keep the discussion all here.

**Comment by bondydaa** [comment]
threw out the idea to the Callstack team here https://expensify.slack.com/archives/C03UK30EA1Z/p1677274105842439

**Comment by JmillsExpensify** [comment]
@bondydaa What's going to be the process for compensating any issue reporters on the linked issues you closed?

**Comment by bondydaa** [comment]
oh hmm 🤷 no idea. 

First thought is maybe we just pay them now? they are bugs that will be fixed eventually. 

I guess the problem could be if people keep reporting the same thing over and over but I guess we just need to be diligent about making sure only new bugs that haven't already been reported get added to this issue?

**Comment by JmillsExpensify** [comment]
I think you should:

 - Open them back up
 - Make sure they're on HOLD
 - Put a `monthly` label on them

**Comment by JmillsExpensify** [comment]
At that point, the most bumps someone could get from Melvin would be twelve times a year. Good news is that this project should last much much less longer than that, as this is a pretty critical part of the app, and it's broken right now. I'd expect us to resolve everything much sooner, basically in single digit months.

**Comment by bondydaa** [comment]
👍 sounds good

**Comment by gedu** [comment]
Hey, can I take this one?

**Comment by gedu** [comment]
I keep looking into this, wrote down some use cases and main flows, and looking into best way to handle this so I can write a proposal

**Comment by MonilBhavsar** [comment]
Prepared a predesign draft. We can discuss and then post in the channel. https://docs.google.com/document/d/1laqFNEmt4P9MminYKv5qGwvlR5rBNTbY1zMnMGImKPg/edit

**Comment by gedu** [comment]
Working on use cases, some talks are in [slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1678190130092049)

**Comment by MonilBhavsar** [comment]
Posted a predesign here https://expensify.slack.com/archives/C01GTK53T8Q/p1678250769318789

**Comment by gedu** [comment]
Preliminar thread about the use cases and if we agree with them are in this [thread on slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1678250769318789)

**Comment by gedu** [comment]
We are working internally to determine the optimal approach for the development phase and the most suitable architecture to use.

**Comment by gedu** [comment]
A new [doc was created](https://docs.google.com/document/d/1ptG18QEpBmunt1wCFuCVfFA9pk0Z_zrcsM54p6KoCMs/edit?usp=sharing) so with @MonilBhavsar we can work on a proposal

**Comment by MonilBhavsar** [comment]
To ensure the proposal works fine and fixes all the use cases we outlined, @gedu and I are working on a POC.

**Comment by gedu** [comment]
I merged both codes and working with @MonilBhavsar to cover the cases in the POC

**Comment by hayata-suenaga** [comment]
added another issue related to the read marker. added "HOLD" to it's title and labeled it "monthly" as suggested [here](https://github.com/Expensify/App/issues/15212#issuecomment-1444606304).

**Comment by gedu** [comment]
@MonilBhavsar is looking into the server side to fix some new cases

**Comment by JmillsExpensify** [comment]
@MonilBhavsar I thought I'd check in. Are you still working on those service side fixes?

**Comment by trjExpensify** [comment]
Adding another issue related to the unread marker.

[Unread message indicator persists on mobile version despite opening and reacting to message](https://github.com/Expensify/App/issues/17207)

**Comment by MonilBhavsar** [comment]
Thanks!
Catching up on this. Will provide an update this week

**Comment by JmillsExpensify** [comment]
Ok great thanks! I think this one is pretty bad still, it'd be great to figure out where we are currently at.

**Comment by JmillsExpensify** [comment]
That'd be great. I think this is one of the most significant bugs we have in the app. It's very easy to run into these issues in daily use.

**Comment by JmillsExpensify** [comment]
Following up on that last comment, I'm going to add a `daily` label to this issue. Issues with the new marker line keep coming up. Here's the latest example: https://github.com/Expensify/App/issues/18180.

**Comment by JmillsExpensify** [comment]
Here's another instance: https://github.com/Expensify/App/issues/16883.

**Comment by MonilBhavsar** [comment]
I am almost done with wave 1. Making this my top priority now

**Comment by JmillsExpensify** [comment]
Thanks! It be great to get this one behind us.

**Comment by MonilBhavsar** [comment]
Working on this and going good.
We started implementing POC a while ago but lot of code has been majorly refactored around report. So we need to start again.
I have linked the PR's.
Will give another update before EOD.

**Comment by JmillsExpensify** [comment]
Thanks @MonilBhavsar! As a heads up, this is now part of the WAQ Everywhere project.

**Comment by MonilBhavsar** [comment]
We have figured out another root cause while working on this. So for this issue right now there are two root causes:

1. We update and return `lastReadTime` in `OpenReport` command and that creates a kind of race condition for `lastReadTime` and as client receives latest time, it calculates false whether or not to show the unread marker. This has been most probably fixed in the draft PR's.

2. We're passing other chat's report actions to calculate the marker. This I think is the reason why it has been flakey and broken. In `ReportScreen.js` we pass `report` and `reportActions` prop down to `ReportActionsView` component. And we calculate the marker in the constructor of `ReportActionsView` component. When switching between chats we pass the reportActions of the previous chat and hence all goes wrong. For example: If I have chat with userA open and chat with userB with unread marker, and when I switch to chatB, we pass reportActions of chat with userA to calculate newMarker.
After some investigation looks like in `ReportScreen.js`, we try to connect to Onyx with the reportID from the route and seems like connects with the old/previous reportID and get those reportActions and pass ahead. This still needs to be fixed.

cc @gedu

**Comment by JmillsExpensify** [comment]
Wow, number 2 is wild. Great work finding that!

**Comment by gedu** [comment]
We fixed conflicts from the main branch, and prettier. Improving number 2 and fixed manually marking the message as unread. Also, I can now test with the changes done on the server by @MonilBhavsar

**Comment by MonilBhavsar** [comment]
In addition to this:

We continued working on POC PR's.
And managed to nearly fix 3 basic scenarios (1. 1:1 Active chat, 2: Switching to new chat, 3: Marking message as unread) by understanding current implementation and making it work according to the ideal way that is - 1. Fetch report/actions from server 2. Calculate marker based on the local(Onyx) data 3. Mark the report as read.




https://github.com/Expensify/App/assets/32012005/05d265c7-aff0-41f2-99e3-26b5aa4e67cd

**Comment by gedu** [comment]
We start working on improving the performance, and looking into some Onyx behaviors

**Comment by gedu** [comment]
Testing the lastest changes to make sure that all keeps working

**Comment by MonilBhavsar** [comment]
Testing client changes.

**Comment by MonilBhavsar** [comment]
We're going through all the use cases we've listed [here](https://docs.google.com/document/d/1j1UKjl_zVmS_EVVtpeH2ugrlt6INuWKlnp3X_o_WXRY/edit#bookmark=id.2cs4566uf99l) in the doc before sending it out

**Comment by MonilBhavsar** [comment]
We shared proposal in the slack here https://expensify.slack.com/archives/C01GTK53T8Q/p1684836449852569

**Comment by MonilBhavsar** [comment]
https://drive.google.com/file/d/1HFN6YFJRnWq2vZSLbixpiwLWZj8wgSnA/view?usp=share_link

**Comment by MonilBhavsar** [comment]
Proposal is being reviewed and we're discussing on slack

**Comment by MonilBhavsar** [comment]
We got reviews on the proposal. In parallel we have started drafting the doc. https://docs.google.com/document/d/144r92kq3lvJlT2sg9wQGzFHdq5PPyai4eCs5jsBWrZw/edit#

**Comment by MonilBhavsar** [comment]
We're working on a doc now.

**Comment by MonilBhavsar** [comment]
Working through the process

**Comment by MitchExpensify** [comment]
Based on [this Slack thread](https://expensify.slack.com/archives/C049HHMV9SM/p1685721982368219?thread_ts=1685659497.038729&cid=C049HHMV9SM) is sounds like we want to add this issue to this tracking, is that right? [[HOLD] Archived announce and admin channels dropped to bottom of the list with unread indicator](https://github.com/Expensify/App/issues/20148#top)

**Comment by MonilBhavsar** [comment]
Yes, let's put this on HOLD for this issue. It is not directly related to unread marker, but involves marking message as read. so should be fixed/updated in this issue

**Comment by MitchExpensify** [comment]
Awesome, thanks for clarifying!

**Comment by MonilBhavsar** [comment]
All good! Thanks for flagging this.

**Comment by MonilBhavsar** [comment]
Update: We shared whatsnext proposal here https://expensify.slack.com/archives/CC7NECV4L/p1686749939851689

**Comment by MonilBhavsar** [comment]
Sent an email to strategy

**Comment by joekaufmanexpensify** [comment]
@MonilBhavsar we were discussing here in Slack that [this issue](https://github.com/Expensify/App/issues/17588) might be fixed as a result of this project. Curious if you think that's the case?

**Comment by MonilBhavsar** [comment]
Looks similar to this [issue](https://github.com/Expensify/App/issues/15212#issuecomment-1589606728). We can put it on HOLD for this issue, as we'll be refactoring some code around this.

**Comment by joekaufmanexpensify** [comment]
Sounds good! I'll put it on hold for this one, and we can retest to see if this is fixed after this project is done.

**Comment by mountiny** [comment]
I have read and reviewed this Design Doc!

**Comment by puneetlath** [comment]
I have read and reviewed this Design Doc!

**Comment by quinthar** [comment]
I have read and reviewed this Design Doc!

**Comment by RachCHopkins** [comment]
I have read and reviewed this Design Doc!

**Comment by saracouto** [comment]
I have read and reviewed this Design Doc!

**Comment by pecanoro** [comment]
I have read and reviewed this Design Doc!

**Comment by kortneywynn** [comment]
I have read and reviewed this Design Doc!

**Comment by Gonals** [comment]
I have read and reviewed this Design Doc!

**Comment by robertjchen** [comment]
I have read and reviewed this Design Doc!

**Comment by JmillsExpensify** [comment]
@MonilBhavsar Are we good to close out the HL reviews yet?

**Comment by mallenexpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by mallenexpensify** [comment]
Looks like it has all the necessary review too
<img width="727" alt="image" src="https://github.com/Expensify/App/assets/22508304/1ecafb0d-f3b3-4b2a-8421-737838d2f81b">

**Comment by MonilBhavsar** [comment]
Thank you everyone for the review 🙇 
Mail sent and working on detailed section now

**Comment by MonilBhavsar** [comment]
Working on technical section of the doc

**Comment by cead22** [comment]
I have read and reviewed this Design Doc!

**Comment by MonilBhavsar** [comment]
Technical section is out for the review

**Comment by MonilBhavsar** [comment]
Doc is in review

**Comment by Gonals** [comment]
I have read and reviewed this Design Doc!

**Comment by isabelastisser** [comment]
I have read and reviewed this Design Doc!

**Comment by danielrvidal** [comment]
I have read and reviewed this Design Doc!

**Comment by Julesssss** [comment]
I have read and reviewed this Design Doc!

**Comment by Julesssss** [comment]
I have read and reviewed this Design Doc!

**Comment by saracouto** [comment]
I have read and reviewed this Design Doc!

**Comment by twisterdotcom** [comment]
I have read and reviewed this Design Doc!

**Comment by mananjadhav** [comment]
@JmillsExpensify @MonilBhavsar I just came across this issue when I posted on the [slack thread](https://expensify.slack.com/archives/C02NK2DQWUX/p1689097248560249). 

I've been tagged to an [issue](https://github.com/Expensify/App/issues/21115) related to new line marker inconsistency when the sender requests money. does it make sense to add it to the list of issues and put it on hold?

cc - @greg-schroeder

**Comment by AndrewGable** [comment]
I have read and reviewed this Design Doc!

**Comment by kadiealexander** [comment]
I have read and reviewed this Design Doc!

**Comment by grgia** [comment]
I have read and reviewed this Design Doc!

**Comment by cead22** [comment]
I have read and reviewed this Design Doc!

**Comment by stephanieelliott** [comment]
I have read and reviewed this Design Doc!

**Comment by robertjchen** [comment]
I have read and reviewed this Design Doc!

**Comment by MonilBhavsar** [comment]
We've required reviews and we're implementing now 🚀

**Comment by JmillsExpensify** [comment]
I have read and reviewed this Design Doc!

**Comment by JmillsExpensify** [comment]
@MonilBhavsar mind providing an overall update on where we're at?

**Comment by MonilBhavsar** [comment]
Updates in respective issues. We're have got PR's up. Testing various cases

**Comment by JmillsExpensify** [comment]
Woo! This is exciting.

**Comment by MonilBhavsar** [comment]
We're making progress and PR's are in review

**Comment by MonilBhavsar** [comment]
Same ^ Going through reviews

**Comment by MonilBhavsar** [comment]
Close to merge the mega App PR

**Comment by JmillsExpensify** [comment]
What's the latest? Are we still close?

**Comment by MonilBhavsar** [comment]
The App PR was deployed. We're seeing flaky behavior sometimes where marker suddenly disappears. Working on that on issue here https://github.com/Expensify/App/issues/23171

**Comment by MonilBhavsar** [comment]
Followup issue in progress https://github.com/Expensify/App/issues/25673

**Comment by neil-marcellini** [comment]
@MonilBhavsar will this issue fix https://github.com/Expensify/App/issues/16883?

**Comment by MonilBhavsar** [comment]
@neil-marcellini yes, that issue will be fixed.
 ---
 Backend PR is in review now

**Comment by MonilBhavsar** [comment]
Waiting for https://github.com/Expensify/Web-Expensify/pull/37324 to be deployed

**Comment by MonilBhavsar** [comment]
Implementation issues are deployed. Working on a known flaky issue here https://github.com/Expensify/App/issues/25673

**Comment by MonilBhavsar** [comment]
PR is being reviewed https://github.com/Expensify/App/pull/25935

**Comment by JmillsExpensify** [comment]
Looks like conflicts were recently resolved and this is ready for another PR review.

**Comment by JmillsExpensify** [comment]
@MonilBhavsar Once we get this PR merged, what's the plan for tackling the linked issues above? Shall we have the issue reporter re-test to see if it still exists and open up to the community for assignments if so?

**Comment by MonilBhavsar** [comment]
The plan is to comment in the linked github issues that are on HOLD for this issue, retest and close if fixed. Else evaluate and we'll fix it!

**Comment by JmillsExpensify** [comment]
Ok great. I would love to wrap this one up by the end of the year. We're approaching a year old at that point.

**Comment by JmillsExpensify** [comment]
Looks like about half of the linked issues are still outstanding.

**Comment by neil-marcellini** [comment]
@JmillsExpensify @MonilBhavsar quoting from what I [said on this held issue](https://github.com/Expensify/App/issues/17692#issuecomment-1819640652).

> We're holding on [[Tracking] Redesign Unread Markers](https://github.com/Expensify/App/issues/15212#top) where there is essentially no progress. I think we really need to bump up the priority of that project since it's fundamental to NewDot working properly. Maybe we could make it part of #vip-split-p2p-chat?

**Comment by JmillsExpensify** [comment]
Agree we should wrap this up. Can we get the remaining issues addressed by the end of the year, basically before Christmas hits?

**Comment by JmillsExpensify** [comment]
Still working through the remaining issues.

**Comment by neil-marcellini** [comment]
Also this one [[$1000] Chat - The group chat status does not show unread after the user goes online](https://github.com/Expensify/App/issues/17692#top)

**Comment by JmillsExpensify** [comment]
@MonilBhavsar Do you think we'll be through these by year end by chance?

**Comment by MonilBhavsar** [comment]
Yes, going pretty good and we're close

1. ~Marker displayed chaotically https://github.com/Expensify/App/issues/27456~ - Merged!
2. ~Inconsistency for Money request and split previews https://github.com/Expensify/App/issues/29376~ - Merged!
3. Flaky behavior while deleting a message https://github.com/Expensify/App/issues/31637 - WIP
4. ~Inconsistency when navigating between threads https://github.com/Expensify/App/issues/31026~ - Fixed by 1  and closed
5. Unread issue in offline mode https://github.com/Expensify/App/issues/17692 - WIP, PR's are open

**Comment by JmillsExpensify** [comment]
Nice, solid progress!

**Comment by JmillsExpensify** [comment]
Looks like we still have a couple left in the list.

**Comment by MonilBhavsar** [comment]
We can close this after this WIP bug report is closed https://github.com/Expensify/App/issues/31714

**Comment by MonilBhavsar** [comment]
Working on the bug https://github.com/Expensify/App/issues/31714 along with waves

**Comment by JmillsExpensify** [comment]
What's the latest on this initiative?

**Comment by MonilBhavsar** [comment]
Still trying to close the last bug https://github.com/Expensify/App/issues/31714 amidst wave issues

**Comment by JmillsExpensify** [comment]
Same as above. Still trying to close that last bug.

**Comment by MonilBhavsar** [comment]
Had a chat with Eduardo from callstack team for the bug and we're working on closing it

**Comment by MonilBhavsar** [comment]
App PR is in review https://github.com/Expensify/App/pull/25771

**Comment by JmillsExpensify** [comment]
That PR is merged right? At that point are we good to close this tracking issue?

**Comment by hayata-suenaga** [comment]
[The PR has been deployed to production](https://github.com/Expensify/App/pull/25771#issuecomment-1999343344)

**Comment by MonilBhavsar** [comment]
Yes, closing and sending an email! 🚀
